src/Juvix/Compiler/Concrete/Translation/FromParsed/Analysis/Scoping.hs

```diff
@@ -2086,16 +2088,70 @@ checkExpressionAtom e = case e of
   AtomList l -> pure . AtomList <$> checkList l
   AtomIterator i -> pure . AtomIterator <$> checkIterator i
   AtomNamedApplication i -> pure . AtomNamedApplication <$> checkNamedApplication i
+  AtomRecordCreation i -> pure . AtomRecordCreation <$> checkRecordCreation i
   AtomRecordUpdate i -> pure . AtomRecordUpdate <$> checkRecordUpdate i
 
+checkRecordCreation :: forall r. (Members '[Error ScoperError, State Scope, State ScoperState, Reader ScopeParameters, InfoTableBuilder, NameIdGen] r) => RecordCreation 'Parsed -> Sem r (RecordCreation 'Scoped)
+checkRecordCreation RecordCreation {..} = do
+  constrIden <- getNameOfKind KNameConstructor _recordCreationConstructor
+  let cname = constrIden ^. scopedIdenFinal
+  tab <- getInfoTable
+  let mci = HashMap.lookup (cname ^. S.nameId) (tab ^. infoConstructors)
+  case mci of
+    Just ci -> do
+      let name = NameUnqualified (ci ^. constructorInfoTypeName . nameConcrete)
+          nameId = ci ^. constructorInfoTypeName . S.nameId
+      info <- getRecordInfo' name nameId
+      let sig = info ^. recordInfoSignature
+      (vars', fields') <- withLocalScope $ localBindings $ ignoreSyntax $ do
+        vs <- mapM (reserveFunctionSymbol . (^. fieldDefineFunDef)) _recordCreationFields
+        fs <- mapM (checkDefineField sig) _recordCreationFields
+        return (vs, fs)
+      let extra' =
+            RecordCreationExtra
+              { _recordCreationExtraSignature = sig,
+                _recordCreationExtraVars = vars'
+              }
+          snames = HashSet.fromList (HashMap.keys (sig ^. recordNames))
+          sfields = HashSet.fromList (map (^. fieldDefineFunDef . signName . nameConcrete) (toList fields'))
+          missingFields = HashSet.difference snames sfields
+      unless (null missingFields) $
+        throw (ErrMissingFields (MissingFields (cname ^. nameConcrete) missingFields))
+      return
+        RecordCreation
+          { _recordCreationConstructor = constrIden,
+            _recordCreationFields = fields',
+            _recordCreationExtra = Irrelevant extra',
+            _recordCreationAtKw
+          }
+    Nothing ->
+      throw (ErrNotAConstructor (NotAConstructor (cname ^. nameConcrete)))
```

Comment:
move the `Nothing` case before the `Just`. Otherwise it is a bit hard to read

Reply:
done